### DISCUSSION TITLE
Possible bug in METcalcpy?

### DISCUSSION THREAD

**Discussion by mollybsmith-noaa** [dtcenter/METplus#2278] [Statistical Computation] [answered]
Hi all,

Not sure if this is a bug or desired behavior, but I've noticed that in METcalcpy, if you're trying to plot bias-corrected RMSE, sometimes the bias correction can make the MSE negative, so that when you take the square root of it you get a NaN back. Is there a method for this stat that applies bias correction *after* the square root is taken, or are there just going to be cases where BCRMSE is NaN?

**Comment by georgemccabe** [comment]
@bikegeek , @TatianaBurek , do you have any insight here?

**Reply by TatianaBurek** [reply]
Here is Tara's answer:
"I'd say at this point no, but we may want to revisit with Eric after this release."
https://dtc-metplus.slack.com/archives/C02P4NVB6HK/p1690922935287339?thread_ts=1690911401.487659&cid=C02P4NVB6HK

**Comment by jprestop** [comment]
@mollybsmith-noaa Do you have any further questions or would it be ok to mark this Discussion as Answered?

**Reply by mollybsmith-noaa** [reply]
No, definitely not answered at this point. I don't think anyone ever roped in Eric. The most I heard was that Tatiana verified that the behavior in METcalcpy was the same as the behavior in the R scripts, but that doesn't mean it's the best way to calculate this statistic? Basically if you have a mean close to zero with a large negative bias you can't really ever use BCRMSE, but you should be able to.

**Reply by jprestop** [reply]
Thank you, @mollybsmith-noaa, for letting us know.  @ericgilleland are you help to help out with this Discussion?

**Reply by mollybsmith-noaa** [reply]
Thanks!

**Reply by ericgilleland** [reply]
Can someone point me to the MET documentation for the bias-corrected RMSE?  I want to see if it is coming from a paper somewhere.  As it is, the only thing I could find, which was what I assumed it would be, is in Jolliffe and Stephenson's book, they talk about bias correcting MSE in sec. 5.4.8.  By their definition, it should not ever be negative as it is essentially the sum of square terms (x - bias - mean(x) )^2.

**Reply by jprestop** [reply]
HI @ericgilleland Thanks for asking.  Searching in the MET documentation, I came across this:
https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html#bias-corrected-mse

Is that helpful?

**Reply by mollybsmith-noaa** [reply]
In METcalcpy, bias-corrected MSE is calculated with:
` mse = calculate_mse(input_data, columns_names, aggregation)`
` me = calculate_me(input_data, columns_names, aggregation)`
` result = mse - me ** 2`
So it can end up negative if MSE is small and bias is large.

**Reply by ericgilleland** [reply]
Thanks, Julie!  Yes, from what I see in our documentation, it is the same bias-corrected RMSE as in Jolliffe and Stephenson.  In our docs, we give it as MSE = ME^2 + s_{f-o}^2, and seems to be based on Wilks' decomposition.  So, MSE - ME ** 2 should be non-negative because MSE - ME ** 2 = s_{f-o}^2=s_f^2 + s_o^2 - 2s_fs_o >= 0.  Are the negative values large or small?  I could see where rounding at different stages would cause small negative values.

**Reply by mollybsmith-noaa** [reply]
Thanks for confirming that the function looks correct! The negative values that I've seem so far have been small. `calculate_mse` and `calculate_me` do perform rounding operations, so that might be what is leading to this. We need to prevent the errors, though, so what would be the best way to solve this programatically?

**Reply by jprestop** [reply]
Thank you @ericgilleland and @mollybsmith-noaa.  Molly, since this is coming from METcalcpy, I am tagging @fisherhucar and @bikegeek who should be able to assist.

**Reply by bikegeek** [reply]
@mollybsmith-noaa , in looking at the code, it looks like the calculate_mse() is in sl1l2_statistics.py- is that correct?

Before I remove the rounding, I'll check with Tatiana to see why it is there, perhaps to be consistent with METviewer???   Naively, I could put a check to see if the value is negative and return the non-rounded version. But I'd need to see if removing the rounding actually addresses this issue.

**Reply by bikegeek** [reply, marked as answer]
Github issue: https://github.com/dtcenter/METcalcpy/issues/329

@mollybsmith-noaa, @ericgilleland et al please feel free to add any other comments to the issue.

**Reply by mollybsmith-noaa** [reply]
@bikegeek That is the correct module in METcalcpy. The issue is that `calculate_bcrmse` throws somewhat frequent errors because `calculate_bcmse` sometimes returns negative numbers.

**Reply by mollybsmith-noaa** [reply]
I wrote this in the issue for this problem, but I want to be clear that the problem is NOT that `calculate_mse` and `calculate_me` are returning negative numbers, they are not doing so. The problem is that the difference between the two can erroneously be negative, which produces an error when trying to take the square root of `calculate_bcmse` to produce `calculate_bcrmse`.

**Comment by bikegeek** [comment]
locking this conversation.  Any additional comments should be directed to the Github Issue:  https://github.com/dtcenter/METcalcpy/issues/329
